### PR TITLE
Limit number of alerts or series produced by a rule

### LINKED ISSUE #9225: Limits on the number of alerts generated by a rule
We have a customer with a faulty rule that generated 30K alerts from a single rule. Now this makes the `/api/v1/rules` API nearly impossible to use with 15MB+ of data. We have scrape time limits on the samples we scrape, but I think we should also have limits on the number of samples each rule evaluation generates.

### PR DESCRIPTION
**Pull request by LeviHarrison**
This PR adds support for a group-level limit of the number of alerts produced by an alerting rule or series produced by a recording rule. If the limit is exceeded, the active alerts are removed if it's an alerting rule, and the rule returns an error which subsequently sets the health to bad.

Fixes #9225

### CODE REVIEW

**Review by dannykopping — COMMENTED** [review]
Looks great @LeviHarrison :+1:
I'd running into a similar issue implementing recording rules for Loki. Would love to see this merged so we can vendor it in.

File: rules/recording_test.go [hunk 1]
```diff
@@ -69,12 +71,41 @@ func TestRuleEval(t *testing.T) {
...
+			err:    "exceeded limit -1 with 1 samples",
+		},
 	}
 
 	for _, test := range suite {
 		rule := NewRecordingRule(test.name, test.expr, test.labels)
-		result, err := rule.Eval(ctx, now, EngineQueryFunc(engine, storage), nil)
-		require.NoError(t, err)
+		result, err := rule.Eval(ctx, now, EngineQueryFunc(engine, storage), nil, test.limit)
+		if err != nil {
```
Comment (roidelapluie) [inline rules/recording_test.go:106]:
```suggestion
		if test.err == nil {
		require.NoError(t, err)
		}
		else {
		require.Equal(t, test.err, err.Error())
		}
```

File: docs/configuration/recording_rules.md [hunk 2]
```diff
@@ -78,6 +78,10 @@ name: <string>
 # How often rules in the group are evaluated.
 [ interval: <duration> | default = global.evaluation_interval ]
 
+# The limit of alerts an alerting rule can produce, or the limit of
```
Comment (roidelapluie) [inline docs/configuration/recording_rules.md:81]:
```suggestion
# Limit the number of alerts and series individual rules can produce.
# This is scoped per rule.
```

File: rules/alerting.go [hunk 3]
```diff
@@ -414,6 +414,16 @@ func (r *AlertingRule) Eval(ctx context.Context, ts time.Time, query QueryFunc,
 		}
 	}
 
+	numActive := len(r.active)
+	if limit != 0 && numActive > limit {
+		r.active = map[uint64]*Alert{}
+		return nil, errors.Errorf("exceeded limit of %v with %v alerts", limit, numActive)
```
Comment (roidelapluie) [inline rules/alerting.go:420]: we tend to avoid %v

File: rules/alerting.go [hunk 4]
```diff
@@ -414,6 +414,16 @@ func (r *AlertingRule) Eval(ctx context.Context, ts time.Time, query QueryFunc,
...
+	}
+
+	// We have already acquired the lock above hence using SetHealth and
+	// SetLastError will deadlock.
+	r.health = HealthGood
```
Comment (roidelapluie) [inline rules/alerting.go:425]: I don't think this is needed anymore

File: rules/manager.go [hunk 5]
```diff
@@ -850,6 +856,10 @@ func (g *Group) Equals(ng *Group) bool {
 		return false
 	}
 
+	if g.limit != ng.limit {
+		return false
```
Comment (roidelapluie) [inline rules/manager.go:860]: do we test this?

File: docs/configuration/recording_rules.md (hunk 2 above)
Comment (LeviHarrison) [reply docs/configuration/recording_rules.md:81]: The limit is applied to individual rules, but it's configured for the entire rule group, which makes "This is scoped per rule." confusing.

Comment (roidelapluie) [reply docs/configuration/recording_rules.md:81]: yeah we can omit that sentence. but the first one should make it clear that it is per rule.

File: rules/manager.go (hunk 5 above)
Comment (LeviHarrison) [reply rules/manager.go:860]: Yes https://github.com/prometheus/prometheus/blob/e5c8575188f00dc7f6a53b884828559efebb5011/rules/manager_test.go#L776-L780

File: docs/configuration/recording_rules.md (hunk 2 above)
Comment (LeviHarrison) [reply docs/configuration/recording_rules.md]: Although I think it's possible that an alerting rule can produce more series than alerts, and we only limit alerts. Should we also limit series?

Comment (roidelapluie) [reply docs/configuration/recording_rules.md]: yes we should limit series

**Comment by dannykopping** [comment]
@roidelapluie would you mind giving this another review / merge?
We'd love to vendor this into Loki as well.

**Comment by roidelapluie** [comment]
The code looks fine, but I need to take this in real situations to really assess the behavior end-to-end. I'll do it asap.

**Comment by roidelapluie** [comment]
I have run some tests, it works as expected.  One thing surprised me: we do not report evaluation errors in the /alerts page, only in the /rules page.

I wonder if we should add more documentation. With this setting, an alert can simply never fire if it directly hit the threshold. This is reflected on prometheus_rule_evaluation_failures_total, which we could mention as well.

**Comment by LeviHarrison** [comment]
Thanks for testing :)

**Comment by roidelapluie** [comment]
Note to self: we should also note that after getting back below the limit, the alert need to go via the for: duration again.

**Comment by roidelapluie** [comment]
Mmh this pull request takes into consideration the "inactive alerts", which are not pending and not firing. I feel like they should not count, WDYT?

**Comment by aknuds1** [comment]
Is there a reason for introducing a single parameter limiting both the number of alerts and series a rule can generate? For context, we would like to limit just the number of alerts a rule can generate in Grafana Mimir. Would it make sense if we add a new parameter to limit only the number of alerts the rules of a rule group can generate, and potentially a corresponding one limiting the number of of series?

**Pull request merged**